File: cp_nim/src/bin/abc428-b-most-frequent-substrings.nim
# Created by Ayush Biswas at 2026/07/18 18:17
# https://atcoder.jp/contests/adt_all_20260623_2/tasks/abc428_b
import ../lib/cpio

# @code begin

import strutils, tables, sugar, sequtils, algorithm

var r = newReader()

let
  (n, k) = r[int, int]
  s = r[string]

var freqs = initCountTable[string]()
for i in 0 .. n - k:
  freqs.inc(s[i..<i+k])

let maxFreq = freqs.values.toSeq.max()
echo maxFreq
let res = collect:
  for k, v in freqs:
    if v == maxFreq: k
echo res.sorted.join(" ")

# @code end

File: cp_rust/src/bin/abc428-b-most-frequent-substrings.rs
// Created by Ayush Biswas at 2026/07/07 11:52
// https://atcoder.jp/contests/adt_all_20260623_2/tasks/abc428_b
use cp_lib::*;

// @code begin
use cpio::*;
use std::{collections::HashMap, ops::AddAssign};

sol! {
    fn solution(
        [n, k]: [usize; 2],
        s: String
    ) -> Lines<Words<String>> {
        let mut occ: HashMap<String, usize> = HashMap::new();
        for i in 0..=n-k {
            let substr = s.get(i..i+k).unwrap().to_string();
            occ.entry(substr).or_insert(0).add_assign(1);
        }

        let max_count = *occ.values().max().unwrap();
        let mut res = vec![];
        for (k, v) in occ {
            if v == max_count {
                res.push(k);
            }
        }
        res.sort();

        vec![
            vec![max_count.to_string()].into(),
            res.into()
        ].into()
    }
}
// @code end
